[CSS] Allow declarations inside @scope rule without style rule ancestor

mdubet opened the pull request:
#### c6177048cd501236c4297419b0597518b490f586
<pre>
[CSS] Allow declarations inside @scope rule without style rule ancestor
<a href="https://bugs.webkit.org/show_bug.cgi?id=287909">https://bugs.webkit.org/show_bug.cgi?id=287909</a>
<a href="https://rdar.apple.com/145101643">rdar://145101643</a>

Reviewed by Antti Koivisto.

<a href="https://github.com/w3c/csswg-drafts/issues/10389">https://github.com/w3c/csswg-drafts/issues/10389</a>

<a href="https://drafts.csswg.org/css-cascade-6/#scope-scope">https://drafts.csswg.org/css-cascade-6/#scope-scope</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-declarations-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-nesting-expected.txt:
* Source/WebCore/css/CSSGroupingRule.cpp:
(WebCore::CSSGroupingRule::insertRule):
* Source/WebCore/css/StyleSheetContents.cpp:
(WebCore::StyleSheetContents::hasNestingRules const):
* Source/WebCore/css/parser/CSSParser.cpp:
(WebCore::CSSParser::consumeNestedGroupRules):
(WebCore::CSSParser::consumeBlockContent):
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addStyleRule):

Canonical link: <a href="https://commits.webkit.org/296262@main">https://commits.webkit.org/296262@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b9ffe0a23350f8ad07f064f01d76cba1ca33020f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/107551 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27234 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17644 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/112768 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58093 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/109515 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27925 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/35735 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81655 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110480 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22134 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96940 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62036 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21571 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15079 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/57531 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91504 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15113 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/115869 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34619 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25530 "Found 3 new test failures: inspector/model/remote-object-get-properties.html inspector/runtime/getDisplayableProperties.html inspector/runtime/getProperties.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90691 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/34995 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93191 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90432 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23137 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35350 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13130 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/30382 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/34540 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40096 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34286 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37646 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/35949 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/css/CSSGroupingRule.cpp:
```diff
@@ -97,11 +97,15 @@ ExceptionOr<unsigned> CSSGroupingRule::insertRule(const String& ruleString, unsi
         return Exception { ExceptionCode::HierarchyRequestError };
     }
 
+    // FIXME: Support inserting rule inside a @scope rule without style rule ancestor from CSSOM
```
Please add a Bugzilla number here so we remember to do this.

brentfulgham replied:
i.e., `FIXME(123456): Support inserting ...`

brentfulgham commented on Source/WebCore/css/StyleSheetContents.cpp:
```diff
@@ -504,7 +504,7 @@ Document* StyleSheetContents::singleOwnerDocument() const
 static bool traverseRulesInVector(const Vector<Ref<StyleRuleBase>>& rules, NOESCAPE const Function<bool(const StyleRuleBase&)>& handler)
 {
-    for (auto& rule : rules) {
+    for (const auto& rule : rules) {
```
Why is this change necessary?

nt1m replied:
We don't do const variables in WebKit

mdubet replied:
It's not necessary, just marking the ref as const for more security

brentfulgham commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -1486,7 +1488,16 @@ void CSSParserImpl::consumeBlockContent(CSSParserTokenRange range, StyleRuleType
+                    }
+                    return false;
+                }();
+                if (hasAncestorStyleRule && !rule->isGroupRule())
```
Is this a hot code path? Maybe it would be faster to always check `!rule->isGroupRule()` before iterating over the rule type stack since we can fast-fail that way?

mdubet replied:
Good idea it could be potentially hot. Changing the code to use short-circuiting

brentfulgham commented on Source/WebCore/style/RuleSetBuilder.cpp:
```diff
@@ -337,10 +337,15 @@ void RuleSetBuilder::addStyleRule(StyleRuleNestedDeclarations& rule)
-        if (m_ancestorStack.last() == CSSParserEnum::NestedContextType::Style)
+
+        // Parent is a style rule.
+        if (m_ancestorStack.last() == CSSParserEnum::NestedContextType::Style) {
```
```
inline bool RuleSetBuilder::parentIsStyleRule()
{
    m_ancestorStack.last() == CSSParserEnum::NestedContextType::Style;
}

inline bool RuleSetBuilder::parentIsScopeRule()
{
    m_ancestorStack.last() == CSSParserEnum::NestedContextType:: Scope;
}
```

Then you can get rid of the comment...

mdubet committed:
[CSS] Allow declarations inside @scope rule without style rule ancestor
https://bugs.webkit.org/show_bug.cgi?id=287909
rdar://145101643

Reviewed by Antti Koivisto.

https://github.com/w3c/csswg-drafts/issues/10389

https://drafts.csswg.org/css-cascade-6/#scope-scope

* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-declarations-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-nesting-expected.txt:
* Source/WebCore/css/CSSGroupingRule.cpp:
(WebCore::CSSGroupingRule::insertRule):
* Source/WebCore/css/StyleSheetContents.cpp:
(WebCore::StyleSheetContents::hasNestingRules const):
* Source/WebCore/css/parser/CSSParser.cpp:
(WebCore::CSSParser::consumeNestedGroupRules):
(WebCore::CSSParser::consumeBlockContent):
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addStyleRule):

Canonical link: https://commits.webkit.org/296262@main
